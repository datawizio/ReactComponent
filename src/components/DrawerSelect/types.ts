import React from "react";
import { SelectProps as AntSelectProps } from "./antd/AntSelect";
import { SelectValue } from "antd/lib/tree-select";
import { AntTreeNode } from "antd/lib/tree";

export interface DrawerSelectProps<VT>
  extends Omit<AntSelectProps<VT>, "onChange"> {
  additionalFilters?: any;
  /**
   * Данные будут загружаться ассинхронно. Будет вызываться функция `loadData`
   */
  asyncData?: boolean;

  hideSearch?: boolean;

  /**
   * Title Drawer-а
   */
  drawerTitle?: string;

  /**
   * Drawer width in px
   */
  drawerWidth?: number;

  /**
   * Label prop
   */
  labelProp?: string;

  /**
   * Функция которая будет вызываться для подгрузки данных с параметрами `searchValue`, `page`
   */
  loadData?: (
    filters: any,
    page: number,
    search: string
  ) => Promise<{ data: any; totalPages: number }>;

  /**
   * Function for customized options
   * */
  optionRender?: (option: any) => any;

  noticeRender?: React.ReactElement | null;

  customFields?: React.ReactElement | null;

  customFieldsHeight?: number;

  multiple?: boolean;

  /**
   * Value prop
   */
  valueProp?: string;

  /**
   * max selected count
   */

  maxSelectedCount?: number;

  maxTagLength?: number;

  /**
   * Подгрузка ассинхронных данных с пагинацией
   */
  withPagination?: boolean;

  /**
   * Event when user clicks Submit
   */
  onChange?: (values: SelectValue, selected?: AntTreeNode) => void;

  onCheckSelectedValue?: (values: SelectValue) => void;

  valueToUncheck?: string | number;

  onLoadData?: (data: any, value: any) => { value?: any };

  onDrawerCancelCallback?: () => void;

  onDrawerSubmitCallback?: () => void;
}