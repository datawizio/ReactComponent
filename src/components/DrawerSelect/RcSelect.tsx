/**
 * To match accessibility requirement, we always provide an input in the component.
 * Other element will not set `tabIndex` to avoid `onBlur` sequence problem.
 * For focused select, we set `aria-live="polite"` to update the accessibility content.
 *
 * ref:
 * - keyboard: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/listbox_role#Keyboard_interactions
 *
 * New api:
 * - listHeight
 * - listItemHeight
 * - component
 *
 * Remove deprecated api:
 * - multiple
 * - tags
 * - combobox
 * - firstActiveValue
 * - dropdownMenuStyle
 * - openClassName (Not list in api)
 *
 * Update:
 * - `backfill` only support `combobox` mode
 * - `combobox` mode not support `labelInValue` since it's meaningless
 * - `getInputElement` only support `combobox` mode
 * - `onChange` return OptionData instead of ReactNode
 * - `filterOption` `onChange` `onSelect` accept OptionData instead of ReactNode
 * - `combobox` mode trigger `onChange` will get `undefined` if no `value` match in Option
 * - `combobox` mode not support `optionLabelProp`
 */

import React from "react";
import { OptionsType as SelectOptionsType } from "rc-select/es/interface";
import SelectOptionList from "rc-select/es/OptionList";
import Option from "rc-select/es/Option";
import OptGroup from "rc-select/es/OptGroup";
import { convertChildrenToData as convertSelectChildrenToData } from "rc-select/es/utils/legacyUtil";
import {
  getLabeledValue as getSelectLabeledValue,
  filterOptions as selectDefaultFilterOptions,
  isValueDisabled as isSelectValueDisabled,
  findValueOption as findSelectValueOption,
  flattenOptions,
  fillOptionsWithMissingValue
} from "rc-select/es/utils/valueUtil";
import generateSelector, { SelectProps } from "../DrawerTreeSelect/RcSelect";
import { RefSelectProps } from "rc-select/es/generate";
import { DefaultValueType } from "rc-select/es/interface/generator";
import warningProps from "rc-select/es/utils/warningPropsUtil";

const RefSelect = generateSelector<SelectOptionsType>({
  prefixCls: "rc-select",
  components: {
    optionList: SelectOptionList
  },
  convertChildrenToData: convertSelectChildrenToData,
  flattenOptions,
  getLabeledValue: getSelectLabeledValue,
  filterOptions: selectDefaultFilterOptions,
  isValueDisabled: isSelectValueDisabled,
  findValueOption: findSelectValueOption,
  warningProps,
  fillOptionsWithMissingValue
});

export type ExportedSelectProps<
  ValueType extends DefaultValueType = DefaultValueType
> = SelectProps<SelectOptionsType, ValueType>;

/**
 * Typescript not support generic with function component,
 * we have to wrap an class component to handle this.
 */
class Select<VT> extends React.Component<SelectProps<SelectOptionsType, VT>> {
  static Option: typeof Option = Option;

  static OptGroup: typeof OptGroup = OptGroup;

  selectRef = React.createRef<RefSelectProps>();

  focus = () => {
    this.selectRef.current.focus();
  };

  blur = () => {
    this.selectRef.current.blur();
  };

  render() {
    return <RefSelect ref={this.selectRef} {...this.props} />;
  }
}

export default Select;
