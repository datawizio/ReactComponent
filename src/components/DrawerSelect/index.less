@import "~antd/es/select/style/index";
@import "~antd/es/tree-select/style/index";
@import "~antd/es/skeleton/style/index";
@import "~antd/es/style/themes/default.less";
@import "../../styles/variables";
@import "./rtl";

.drawer-select {
  .ant-select-selector {
    & > span {
      overflow: hidden;
    }
  }

  .ant-select-selection-search-input {
    font-size: 1px;
  }

  .ant-select-selection-item {
    .ant-select-selection-item-content {
      border: 0;
      background: transparent;
      margin: 0;
      padding: 0;
      font-size: 14px;

      .anticon-close {
        font-size: 12px;
      }
    }
  }

  span {
    & > .ant-select-selection-placeholder {
      display: none;
    }

    &:first-of-type > .ant-select-selection-placeholder {
      display: block;
    }
  }
}

.drawer-select-dropdown-fake.ant-select-dropdown.ant-select-dropdown {
  visibility: hidden;
}

.drawer-select-dropdown {
  .ant-drawer-body {
    overflow: hidden;
    padding-bottom: @padding-lg - 16px;
    display: flex;
    flex-direction: column;

    .drawer-select-field {
      margin-bottom: 12px;
    }
  }

  .ant-input-search {
    margin-bottom: 10px;
  }

  .ant-select-item {
    padding-left: 0;
    align-items: center;
    transition: all 0.3s;

    &::before {
      content: " ";
      margin-right: 10px;
      min-width: @checkbox-size;
      width: @checkbox-size;
      height: @checkbox-size;
      border: @checkbox-border-width @border-style-base @border-color-base;
      border-radius: @border-radius-base;
    }

    &:hover {
      &::before {
        border-color: @checkbox-color;
      }
    }

    &.ant-select-item-option-selected {
      &::before {
        background-color: @checkbox-color;
        border-color: @checkbox-color;
      }

      &::after {
        @check-width: (@checkbox-size / 14) * 5px;
        @check-height: (@checkbox-size / 14) * 8px;

        position: absolute;
        display: table;
        border: 2px solid @checkbox-check-color;
        border-top: 0;
        border-left: 0;
        transform: rotate(45deg) scale(1) translate(-50%, -50%);
        opacity: 1;
        transition: all 0.2s @ease-out-back 0.1s;
        content: " ";

        top: 15px;
        left: 4px;

        width: @check-width;
        height: @check-height;
      }
    }

    &.ant-select-item-option-active,
    &.ant-select-item-option-selected {
      background-color: transparent;
      font-weight: 400;
    }

    .ant-select-item-option-state {
      display: none;
    }
  }

  .drawer-select-selected {
    padding-top: 5px;
    display: flex;
    background: inherit;
    color: @drawer-selected-text-color;

    .drawer-select-selected-title {
      flex: 1;
    }
  }

  .drawer-select-loader-container {
    flex: 1;
    height: 16px;
    overflow: visible;

    .ant-skeleton-title {
      margin: 0;
    }

    .ant-skeleton-content .ant-skeleton-title + .ant-skeleton-paragraph {
      margin-top: 5px;

      li {
        width: 300px !important;
      }
    }
  }
}
