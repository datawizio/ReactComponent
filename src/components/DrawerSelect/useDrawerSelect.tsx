import { useReducer } from "react";

import { SelectValue } from "../DrawerSelect/antd/AntSelect";
import { AntTreeNode } from "antd/es/tree";

export interface IUseDrawerSelect {
  internalLoading: boolean;
  page: number;
  totalPages: number;
  drawerVisible: boolean;
  searchValue: string;
  internalValue: SelectValue;
  selected: AntTreeNode;
  optionsState: any;
}

function reducer(state: IUseDrawerSelect, action: any) {
  switch (action.type) {
    case "remoteLoadDataStart": {
      return { ...state, internalLoading: true, ...action.payload };
    }
    case "remoteLoadDataStop": {
      return { ...state, internalLoading: false, ...action.payload };
    }
    case "drawerCancel": {
      return {
        ...state,
        searchValue: "",
        drawerVisible: false,
        ...action.payload
      };
    }
    case "openDrawer": {
      return {
        ...state,
        drawerVisible: true
      };
    }
    case "drawerSubmit": {
      return {
        ...state,
        searchValue: "",
        drawerVisible: false
      };
    }
    case "setSelected": {
      return {
        ...state,
        selected: action.payload
      };
    }
    case "setSearchValue": {
      return {
        ...state,
        searchValue: action.payload
      };
    }
    case "setInternalValue": {
      return {
        ...state,
        internalValue: action.payload
      };
    }
    case "setState": {
      return {
        ...state,
        ...action.payload
      };
    }
    default:
      throw new Error();
  }
}

export const useDrawerSelect = (initialState: IUseDrawerSelect) => {
  const [state, dispatch] = useReducer<any>(reducer, initialState);

  return [state as IUseDrawerSelect, dispatch as any];
};
