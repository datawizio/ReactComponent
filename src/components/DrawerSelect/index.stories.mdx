import { useState } from "react";
import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import Button from "../Button";
import DrawerSelect from "./index";

<Meta title="Components/Select/DrawerSelect" component={DrawerSelect} />

# DrawerSelect

Компонент `DrawerSelect`. Работает как обычный `Select` только для выбора
открывается `Drawer`.

## Basic

<Preview>
  <Story name="Basic">
    <DrawerSelect
      placeholder="Select"
      showSearch
      allowClear
      options={[
        { title: "1", key: "1", label: "label 1" },
        { title: "12", key: "2", label: "label 2" },
        { title: "123", key: "3", label: "label 3" },
        { title: "1234", key: "4", label: "label 4" }
      ]}
    />
  </Story>
</Preview>

## With pagination

<Preview>
  <Story name="With pagination">
    <DrawerSelect
      placeholder="Select"
      showSearch
      multiple
      allowClear
      withPagination
      loadData={async (search, page) => {
        console.log(search, page);
        if (search === "123") return { data: [], totalPages: 0 };
        console.log(222);
        const response = await fetch("https://randomuser.me/api/?results=1000");
        const json = await response.json();
        const data = json.results.map(user => ({
          title: `${user.name.first} ${user.name.last}`,
          key: user.login.uuid
        }));
        const totalPages = 10;
        return { data, totalPages };
      }}
    />
  </Story>
</Preview>

## Props

<Props of={DrawerSelect} />
