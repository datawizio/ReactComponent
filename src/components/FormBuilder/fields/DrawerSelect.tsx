import React, { useContext } from "react";

import { Form } from "antd";
import { FieldDrawerSelectProps } from "../types";
import ConfigContext from "../../ConfigProvider/context";
import DrawerSelect from "../../DrawerSelect";

export const FieldDrawerSelect: React.FC<FieldDrawerSelectProps> = ({
  label,
  rules,
  name,
  initialValue,
  placeholder,
  multiple,
  options,
  loadData,
  onChange,
  loading,
  ...restProps
}) => {
  const { translate } = useContext(ConfigContext);
  const handleFieldChange = (value: any, selected: any) => {
    console.log(value);
    onChange({
      name,
      value,
      selected
    });
  };

  return (
    <Form.Item
      name={name}
      label={label}
      rules={rules}
      initialValue={initialValue}
    >
      <DrawerSelect
        {...restProps}
        multiple={multiple ? multiple : false}
        options={options}
        drawerTitle={placeholder}
        drawerSearchPlaceholder={translate("SEARCH")}
        submitText={translate("SUBMIT")}
        cancelText={translate("CANCEL")}
        loadingText={translate("LOADING")}
        noDataText={translate("NO_DATA")}
        loadData={loadData}
        loading={loading}
        placeholder={placeholder}
        onChange={handleFieldChange}
        allowClear={true}
      />
    </Form.Item>
  );
};
