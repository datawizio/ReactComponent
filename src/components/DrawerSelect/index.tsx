import React, {
  useRef,
  useCallback,
  useMemo,
  useEffect,
  useContext
} from "react";

import clsx from "clsx";

import AntSelect, { SelectProps as AntSelectProps } from "./antd/AntSelect";
import { SelectValue } from "antd/lib/tree-select";

import SearchInput from "../SearchInput";
import Drawer from "../Drawer";
import Button from "../Button";
import { Skeleton, Tag, message } from "antd";

import { AntTreeNode } from "antd/lib/tree";

import ConfigContext from "../ConfigProvider/context";

import { triggerInputChangeValue } from "../../utils/trigger";
import { getUniqueItems } from "../../utils/data/dataHelpers";
import { useDrawerSelect } from "./useDrawerSelect";

import "./index.less";

export interface DrawerSelectProps<VT>
  extends Omit<AntSelectProps<VT>, "onChange"> {
  additionalFilters?: any;
  /**
   * Данные будут загружаться ассинхронно. Будет вызываться функция `loadData`
   */
  asyncData?: boolean;

  /**
   * Title Drawer-а
   */
  drawerTitle?: string;

  /**
   * Drawer width in px
   */
  drawerWidth?: number;

  /**
   * Label prop
   */
  labelProp?: string;

  /**
   * Функция которая будет вызываться для подгрузки данных с параметрами `searchValue`, `page`
   */
  loadData?: (
    filters: any,
    page: number,
    search: string
  ) => Promise<{ data: any; totalPages: number }>;

  /**
   * Function for customized options
   * */
  optionRender?: (option: any) => any;

  multiple?: boolean;

  /**
   * Value prop
   */
  valueProp?: string;

  /**
   * max selected count
   */

  maxSelectedCount?: number;

  maxTagLength?: number;

  /**
   * Подгрузка ассинхронных данных с пагинацией
   */
  withPagination?: boolean;

  /**
   * Event when user clicks Submit
   */
  onChange?: (values: SelectValue, selected?: AntTreeNode) => void;

  onCheckSelectedValue?: (values: SelectValue) => void;

  valueToUncheck?: string | number;

  onLoadData?: (data: any, value: any) => { value?: any };
}

function extractProperty(array: Array<object>, propertyName: string) {
  return array.map(item => item[propertyName]);
}

function convertOptions(
  source,
  valueProp: string,
  labelProp: string,
  selectedOptions: any[] = [],
  value: string[] = []
) {
  const selected = [];
  const options = [];
  const set = new Set(Array.isArray(value) ? value : [value]);

  const selectedIds = selectedOptions.map(item => item[valueProp]);
  const selectedSet = new Set(selectedIds);

  source.forEach(opt => {
    const obj = {
      ...opt,
      value: opt[valueProp],
      label: opt[labelProp]
    };
    if (selectedSet.has(opt[valueProp])) {
      selected.push(obj);
      return;
    }
    if (set.has(opt[valueProp])) {
      selected.push(obj);
      return;
    }

    options.push(obj);
  });

  return {
    options,
    selected
  };
}

const DrawerSelect: React.FC<DrawerSelectProps<SelectValue>> = props => {
  const {
    additionalFilters,
    asyncData,
    showCheckAll,
    checkAllTitle,
    checkAllKey,
    drawerTitle,
    drawerWidth,
    value,
    isFlatList,
    onChange,
    valueToUncheck,
    onCheckSelectedValue,
    options,
    optionRender,
    optionFilterProp,
    optionLabelProp,
    labelProp,
    loadData,
    loading,
    valueProp,
    mode,
    multiple,
    withPagination,
    maxSelectedCount,
    maxTagLength,
    onLoadData,
    ...restProps
  } = props;

  const { translate } = useContext(ConfigContext);

  const drawerSearchPlaceholder = useMemo(() => translate("SEARCH"), [
    translate
  ]);
  const noDataText = useMemo(() => translate("NO_DATA"), [translate]);
  const loadingText = useMemo(() => translate("LOADING"), [translate]);
  const submitText = useMemo(() => translate("SUBMIT"), [translate]);
  const cancelText = useMemo(() => translate("CANCEL"), [translate]);

  const [
    {
      internalLoading,
      page,
      totalPages,
      drawerVisible,
      searchValue,
      internalValue,
      selected,
      optionsState
    },
    dispatch
  ] = useDrawerSelect({
    internalLoading: loading,
    page: 0,
    totalPages: 1,
    drawerVisible: false,
    searchValue: "",
    internalValue: undefined,
    selected: undefined,
    optionsState: []
  });

  const selectedOptions = useRef<any[]>([]);
  const menuRef = useRef();
  const firstLoadedOptions = useRef<any[]>([]);

  const internalOptions = useMemo(() => {
    return options ? convertOptions(options, valueProp, labelProp).options : [];
  }, [options, valueProp, labelProp]);

  const inputRef = useRef<HTMLInputElement>();

  const setInputRef = (el: HTMLInputElement) => {
    if (el.classList.contains("ant-select-selection-search-input")) {
      inputRef.current = el;
    }
  };

  const triggerOnChange = useCallback(
    value => {
      if (!onChange) return;
      if (!multiple) {
        onChange(value, value ? selected : undefined);
        return;
      }
      onChange(value);
    },
    [selected, multiple, onChange]
  );

  const loadPage = useCallback(
    async (search, page = 0, first = false) => {
      if (!loadData) return;

      if (page !== 0 && page >= totalPages) {
        return;
      }

      let state: any = {};

      if (!page) state.optionsState = [];

      dispatch({
        type: "remoteLoadDataStart",
        payload: state
      });

      const filters = { ...additionalFilters, search };

      filters.selected = extractProperty(selectedOptions.current, valueProp);

      if (first) {
        filters.selected = value;
        filters.first = true;
      }

      const { data, totalPages: pages } = await loadData(filters, page, search);

      if (onLoadData) {
        const { value: loadValue } = onLoadData(data, value);
        if (loadValue) {
          triggerOnChange(loadValue);
        }
      }

      const options = convertOptions(
        data,
        valueProp,
        labelProp,
        first ? [] : selectedOptions.current,
        first ? value : []
      );

      if (first) {
        selectedOptions.current = options.selected;
        firstLoadedOptions.current = options.options;
      }

      state = {
        page: page,
        totalPages: pages
      };

      if (page === 0) {
        state.optionsState = getUniqueItems(
          options.selected.concat(options.options)
        );
      } else {
        state.optionsState = getUniqueItems(
          optionsState.concat(options.options)
        );
      }

      dispatch({
        type: "remoteLoadDataStop",
        payload: state
      });
    },

    //eslint-disable-next-line
    [
      loadData,
      dispatch,
      optionsState,
      valueProp,
      labelProp,
      totalPages,
      onLoadData
    ]
  );

  //  -------- HANDLERS --------
  const closeDrawer = useCallback(() => {
    setTimeout(() => {
      const activeElement = document.activeElement as HTMLElement;
      activeElement.blur();
      try {
        //@ts-ignore
        menuRef.current.ref.current.scrollTo(0);
      } catch (e) {}
    }, 100);
  }, []);

  const handleDrawerCancel = useCallback(() => {
    closeDrawer();

    const payload: any = { internalValue: !multiple && !value ? [] : value };

    if (searchValue && loadData) {
      payload.optionsState = getUniqueItems(
        selectedOptions.current.concat(firstLoadedOptions.current)
      );
    }

    dispatch({
      type: "drawerCancel",
      payload
    });

    if (searchValue) handleSearch("");

    //eslint-disable-next-line
  }, [dispatch, closeDrawer, value, multiple, searchValue, loadData]);

  const handleDrawerSubmit = useCallback(() => {
    closeDrawer();

    const payload: any = {};

    if (searchValue) {
      payload.optionsState = getUniqueItems(
        selectedOptions.current.concat(firstLoadedOptions.current)
      );
    }

    dispatch({
      type: "drawerSubmit",
      payload
    });

    triggerOnChange(internalValue);
    if (searchValue) handleSearch("");

    //eslint-disable-next-line
  }, [dispatch, triggerOnChange, closeDrawer, internalValue, searchValue]);

  const handleDrawerFocus = e => {
    setInputRef(e.target);
    dispatch({
      type: "openDrawer"
    });
    triggerInputChangeValue(inputRef.current, searchValue);
  };

  const handleSearchInputChange = e => {
    const inputValue = e.target.value;
    dispatch({
      type: "setSearchValue",
      payload: inputValue
    });
    triggerInputChangeValue(inputRef.current, inputValue);
  };

  const handleSelectBeforeBlur = useCallback(() => !drawerVisible, [
    drawerVisible
  ]);

  const handleSelect = useCallback(
    (selectedKey, node) => {
      if (selectedKey && onCheckSelectedValue)
        onCheckSelectedValue(selectedKey);
      selectedOptions.current.push(node);
      dispatch({
        type: "setSelected",
        payload: node
      });
    },
    [dispatch, onCheckSelectedValue]
  );

  const handleDeselect = useCallback((_, node) => {
    const index = selectedOptions.current.findIndex(
      option => option.key === node.key
    );
    selectedOptions.current.splice(index, 1);
  }, []);

  const handleSearch = useCallback(
    searchValue => {
      loadPage(searchValue);
    },
    [loadPage]
  );

  const handleChange = useCallback(
    (value, el) => {
      let newValue: any;
      if (multiple) {
        newValue = value;
      } else {
        newValue = value[1] ? [value[1]] : [value[0]];
      }

      if (maxSelectedCount && value.length > maxSelectedCount) {
        const messageKey = "select-over-then-" + maxSelectedCount;
        message.error({
          content: translate("COUNT_MUST_BE_SMALLER_THEN", {
            maxCount: maxSelectedCount
          }),
          key: messageKey
        });

        return;
      }

      dispatch({
        type: "setInternalValue",
        payload: newValue && newValue[0] ? newValue : value
      });

      if (!drawerVisible) {
        triggerOnChange(value);
      }
    },
    [dispatch, drawerVisible, triggerOnChange, multiple, maxSelectedCount]
  );

  const handlePopupScroll = useCallback(
    e => {
      const { target } = e;
      //debounce?
      if (internalLoading) return;
      if (page === totalPages - 1) return;
      if (target.scrollTop + target.offsetHeight >= target.scrollHeight - 100) {
        loadPage(searchValue, page + 1);
      }
    },
    [internalLoading, page, loadPage, searchValue, totalPages]
  );

  // ------- EFFECTS ----------
  useEffect(() => {
    dispatch({
      type: "setInternalValue",
      payload: !multiple && !value ? [] : value
    });
  }, [dispatch, value, multiple]);

  useEffect(() => {
    if (!internalValue || !valueToUncheck) return;
    const payload = internalValue.filter(v => v !== valueToUncheck);
    dispatch({
      type: "setInternalValue",
      payload
    });
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [valueToUncheck]);

  useEffect(() => {
    dispatch({
      type: "setState",
      payload: { internalLoading: loading }
    });
  }, [dispatch, loading]);

  useEffect(() => {
    dispatch({
      type: "setState",
      payload: { optionsState: internalOptions }
    });
  }, [dispatch, internalOptions]);

  useEffect(() => {
    !asyncData && loadData && loadPage("", 0, true);
    return () => {};
    //eslint-disable-next-line
  }, [asyncData, additionalFilters]);

  // -------- RENDERS ---------

  const tagRender = useCallback(
    ({ label, closable, onClose }) => {
      const maxLength = maxTagLength || 20;
      const isLongTag = label?.length > maxLength;

      if (!optionsState || optionsState.length === 0) {
        return (
          <span className="ant-select-selection-placeholder">
            {loadingText}
          </span>
        );
      }

      return (
        <span className="ant-select-selection-item">
          <Tag
            closable={closable}
            onClose={onClose}
            className="ant-select-selection-item-content"
            title={label}
          >
            {isLongTag ? `${label.slice(0, maxLength)}...` : label}
          </Tag>
        </span>
      );
    },
    [optionsState, loadingText]
  );
  const dropdownRender = useCallback(
    menu => {
      menuRef.current = menu;
      return (
        <Drawer
          className={clsx({
            "drawer-select-dropdown": true,
            "drawer-select-dropdown-show-all": showCheckAll
          })}
          title={drawerTitle ? drawerTitle : restProps.placeholder}
          onClose={handleDrawerCancel}
          visible={drawerVisible}
          width={
            window.innerWidth < drawerWidth ? window.innerWidth : drawerWidth
          }
          actions={
            <>
              <Button onClick={handleDrawerCancel}>{cancelText}</Button>
              <Button onClick={handleDrawerSubmit} type="primary">
                {submitText}
              </Button>
            </>
          }
        >
          <SearchInput
            placeholder={drawerSearchPlaceholder}
            value={searchValue}
            onChange={handleSearchInputChange}
            loading={internalLoading}
          />
          {menu}
          <div className="drawer-select-loader-container">
            {internalLoading && (
              <Skeleton
                title={{ width: 300 }}
                paragraph={{ rows: 1 }}
                loading={true}
                active
              />
            )}
          </div>
          {multiple && (
            <div className="drawer-select-selected">
              <div className="drawer-select-selected-title">
                {translate("SELECTED")}
              </div>
              <div className="drawer-select-selected-count">
                {internalValue ? internalValue.length : 0}
              </div>
            </div>
          )}
        </Drawer>
      );
    },
    //eslint-disable-next-line
    [searchValue, internalLoading, drawerVisible, internalValue]
  );

  const properties = {
    filterOption: true,
    ...restProps,
    value: internalValue,
    className: "drawer-select",
    open: drawerVisible,
    loading: internalLoading,
    dropdownRender: dropdownRender,
    searchValue: searchValue,
    tagRender: tagRender,
    dropdownClassName: "drawer-select-dropdown-fake",
    showSearch: true,
    onSearch: handleSearch,
    optionFilterProp: optionFilterProp || "label",
    optionLabelProp: optionLabelProp || "label",
    listHeight: window.innerHeight - 198 - (multiple ? 27 : 0),
    notFoundContent: internalLoading ? loadingText : noDataText,
    onBeforeBlur: handleSelectBeforeBlur,
    onFocus: handleDrawerFocus,
    onSelect: handleSelect,
    onDeselect: handleDeselect,
    onPopupScroll: handlePopupScroll,
    onChange: handleChange
  };

  return optionRender ? (
    <AntSelect {...properties} mode="multiple">
      {optionsState && optionsState.map(option => optionRender(option))}
    </AntSelect>
  ) : (
    <AntSelect {...properties} mode="multiple" options={optionsState} />
  );
};

DrawerSelect.defaultProps = {
  maxTagCount: 10,
  drawerTitle: "",
  drawerWidth: 400,
  labelProp: "title",
  valueProp: "key",
  asyncData: false,
  multiple: false
};

export default DrawerSelect;
