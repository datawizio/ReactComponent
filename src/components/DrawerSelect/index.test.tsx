import React from "react";
import { shallow } from "enzyme";
import DrawerSelect from "./index";
jest.mock("./antd/RcSelect", () => "select");

const setUp = (props?) => shallow(<DrawerSelect />);

describe("DrawerSelect component", () => {
  it("rendered", () => {
    const wrapper = setUp();
  });
});
