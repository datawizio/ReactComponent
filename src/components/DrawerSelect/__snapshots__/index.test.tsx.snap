// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerSelect component Rendered corectly 1`] = `
<antSelect
  className="drawer-select"
  dropdownClassName="drawer-select-dropdown-fake"
  dropdownRender={[Function]}
  filterOption={true}
  listHeight={302}
  maxTagCount={10}
  mode="multiple"
  notFoundContent="NO_DATA"
  onBeforeBlur={[Function]}
  onChange={[Function]}
  onDeselect={[Function]}
  onFocus={[Function]}
  onPopupScroll={[Function]}
  onSearch={[Function]}
  onSelect={[Function]}
  open={false}
  optionFilterProp="label"
  optionLabelProp="label"
  options={Array []}
  searchValue=""
  showSearch={true}
  tagRender={[Function]}
/>
`;
